### PR TITLE
UCP/WIREUP: Remove skipped paths from RMA lane score

### PR DESCRIPTION
**Pull request by tvegas1**
## What
Fix RMA lane score computation when an eligible AM lane was skipped.

## Why ?
The list of lanes must be ordered by score. Without the fix, when `mlx5_0:1` has a better score than `mlx5_2:1`, for put single proto, we erroneously select the later interface.

Broken:
```
lane[0]:  0:dc_mlx5/mlx5_0:1.0 md[0]      -> addr[0].md[0]/ib/sysdev[1] seg 4294967295 amo#0 am am_bw#0
lane[1]:  2:dc_mlx5/mlx5_2:1.0 md[1]      -> addr[2].md[1]/ib/sysdev[3] seg 4294967295 rma_bw#0
lane[2]:  0:dc_mlx5/mlx5_0:1.1 md[0]      -> addr[0].md[0]/ib/sysdev[1] seg 4294967295 rma_bw#1
lane[3]:  2:dc_mlx5/mlx5_2:1.1 md[1]      -> addr[2].md[1]/ib/sysdev[3] seg 4294967295 rma_bw#2
lane[4]:  0:dc_mlx5/mlx5_0:1.2 md[0]      -> addr[0].md[0]/ib/sysdev[1] seg 4294967295 rma_bw#3
```
After fix:
```
lane[0]:  0:dc_mlx5/mlx5_0:1.0 md[0]      -> addr[0].md[0]/ib/sysdev[1] seg 4294967295 am am_bw#0
lane[1]:  0:dc_mlx5/mlx5_0:1.1 md[0]      -> addr[0].md[0]/ib/sysdev[1] seg 4294967295 rma_bw#0
lane[2]:  2:dc_mlx5/mlx5_2:1.0 md[1]      -> addr[2].md[1]/ib/sysdev[3] seg 4294967295 rma_bw#1
lane[3]:  0:dc_mlx5/mlx5_0:1.2 md[0]      -> addr[0].md[0]/ib/sysdev[1] seg 4294967295 rma_bw#2
lane[4]:  2:dc_mlx5/mlx5_2:1.1 md[1]      -> addr[2].md[1]/ib/sysdev[3] seg 4294967295 rma_bw#3
```

## How ?
We need to include the skipped lane in path_index/dev.count.local, to avoid later lane merging with the pure AM one, but we also need to remove it from the RMA score computation keep proper lane ordering.

### Repro
`oshrun -host rock19:1,rock20:1 --map-by ppr:1:node -x UCX_LOG_LEVEL=debug -x UCX_PROTO_INFO=n -x UCX_TLS=dc  -x UCX_NET_DEVICES=mlx5_0:1,mlx5_2:1 -x LD_LIBRARY_PATH ./bench -i 10000 -s 32768 -t quiet -d 32 -k 200 2>&1`

### CODE REVIEW

File: src/ucp/wireup/select.c [hunk 1]
```diff
@@ -1630,8 +1631,13 @@ ucp_wireup_add_bw_lanes(const ucp_wireup_select_params_t *select_params,
         if (num_paths_override &&
             ((skip_dev_index != UCP_NULL_RESOURCE) && /* clang sanitizer */
              (skip_dev_index == dev_index))) {
+            /* Allow path since we skipped one */
             local_num_paths++;
             remote_num_paths++;
+
+            /* Remove that skipped path from score computation */
+            dev_count.local_skip[dev_index]      = 1;
+            dev_count.remote_skip[ae->dev_index] = 1;
```
Comment (brminich) [inline src/ucp/wireup/select.c:1640]: can't we do not increase dev_counts for skipped lane instead?

File: src/ucp/wireup/select.c [hunk 2]
```diff
@@ -1258,14 +1260,15 @@ ucp_wireup_iface_avail_bandwidth(const ucp_worker_iface_t *wiface,
     /* Apply dev num paths ratio after fp8 pack/unpack to make sure it is not
      * neglected because of fp8 inaccuracy
      */
-    local_bw *= ucp_tl_iface_bandwidth_ratio(
-                   context, local_dev_count[dev_index],
-                   wiface->attr.dev_num_paths);
+    count     = dev_count->local[dev_index] - dev_count->local_skip[dev_index],
```
Comment (brminich) [inline src/ucp/wireup/select.c:1263]: in theory `dev_count->local` can be smaller than dev_count->local_skip
maybe add an assert?

File: src/ucp/wireup/select.c (hunk 1 above)
Comment (brminich) [reply src/ucp/wireup/select.c:1640]: ok

**Comment by tvegas1** [comment]
confirmed that perf with this PR matches perf with #9668. still checking atomic use case.

File: src/ucp/wireup/select.c (hunk 2 above)
Comment (tvegas1) [reply src/ucp/wireup/select.c:1275]: added for local and remote

**Comment by tvegas1** [comment]
for runs doing rma/fence/amo in loop, fence ends up taking more time at scale because amo and rma are now separate lanes:
- 2 nodes 2+2 ranks: fence 7usec -> 7.5usec
- 2 nodes 64+64 ranks; fence 261usec -> 342usec

After discussing, we propose to narrow the fix to TAG requesters only, as heavy fence users might not request it.

```
mpirun -x UCX_TLS=dc -x UCX_NET_DEVICES=mlx5_0:1,mlx5_2:1 shmem_all2all_bw -i 10000 -s 32768 -t atomic -d 32 -k 200 
```

**Comment by tvegas1** [comment]
re-tested fix against oshrun and original validation code.

**Review by brminich — APPROVED** [review]

File: src/ucp/wireup/select.c [hunk 3]
```diff
@@ -1875,31 +1893,37 @@ ucp_wireup_add_rma_bw_lanes(const ucp_wireup_select_params_t *select_params,
...
+                                (lane_desc->path_index ==
+                                 UCP_WIREUP_PATH_INDEX_UNDEFINED),
+                        "rsc_index=%u am_lane=%u path_index=%u",
+                        lane_desc->rsc_index, am_lane, lane_desc->path_index);
 
-        lane_desc->path_index = 0;
-        num_paths_override    = 1;
-        break;
+            lane_desc->path_index = 0;
+            num_paths_override    = 1;
```
Comment (yosefe) [inline src/ucp/wireup/select.c:1924]: rename "num_paths_override" to "extra_path" or similar

File: src/ucp/wireup/select.c (hunk 2 above)
Comment (yosefe) [reply src/ucp/wireup/select.c:1275]: rename count to path_index, also in ucp_tl_iface_bandwidth_ratio function

File: src/ucp/wireup/select.c (hunk 1 above)
Comment (yosefe) [reply src/ucp/wireup/select.c:1652]: maybe rename local_skip to first_path_index_local (and remote, accordingly)

File: src/ucp/wireup/select.c (hunk 3 above)
Comment (tvegas1) [reply src/ucp/wireup/select.c]: fixed

**Review by yosefe — APPROVED** [review]

**Pull request merged**